File: go/parameter_type_registry.go
package cucumberexpressions

import (
	"fmt"
	"reflect"
	"regexp"
	"sort"
	"strings"
)

var INTEGER_REGEXPS = []*regexp.Regexp{
	regexp.MustCompile(`-?\d+`),
	regexp.MustCompile(`\d+`),
}
var FLOAT_REGEXPS = []*regexp.Regexp{
	regexp.MustCompile(`[-+]?\d*\.?\d+`),
}
var WORD_REGEXPS = []*regexp.Regexp{
	regexp.MustCompile(`[^\s]+`),
}
var STRING_REGEXPS = []*regexp.Regexp{
	regexp.MustCompile(`"([^"\\]*(\\.[^"\\]*)*)"|'([^'\\]*(\\.[^'\\]*)*)'`),
}
var ANONYMOUS_REGEXPS = `.*`

type ParameterTypeRegistry struct {
	parameterTypeByName    map[string]*ParameterType
	parameterTypesByRegexp map[string][]*ParameterType
	defaultTransformer     ParameterByTypeTransformer
}

func NewParameterTypeRegistry() *ParameterTypeRegistry {
	transformer := BuiltInParameterTransformer{}

	result := &ParameterTypeRegistry{
		parameterTypeByName:    map[string]*ParameterType{},
		parameterTypesByRegexp: map[string][]*ParameterType{},
		defaultTransformer:     transformer,
	}

	intParameterType, err := NewParameterType(
		"int",
		INTEGER_REGEXPS,
		"int",
		func(args ...*string) interface{} {
			i, err := transformer.Transform(*args[0], reflect.Int)
			if err != nil {
				panic(err)
			}
			return i
		},
		true,
		true,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(intParameterType)
	if err != nil {
		panic(err)
	}

	bigintParameterType, err := NewParameterType(
		"biginteger",
		INTEGER_REGEXPS,
		"int",
		func(args ...*string) interface{} {
			i, err := transformer.Transform(*args[0], BigIntKind)
			if err != nil {
				panic(err)
			}
			return i
		},
		false,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(bigintParameterType)
	if err != nil {
		panic(err)
	}

	floatParameterType, err := NewParameterType(
		"float",
		FLOAT_REGEXPS,
		"float32",
		func(args ...*string) interface{} {
			f, err := transformer.Transform(*args[0], reflect.Float32)
			if err != nil {
				panic(err)
			}
			return f
		},
		true,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(floatParameterType)
	if err != nil {
		panic(err)
	}

	doubleParameterType, err := NewParameterType(
		"double",
		FLOAT_REGEXPS,
		"float64",
		func(args ...*string) interface{} {
			f, err := transformer.Transform(*args[0], reflect.Float64)
			if err != nil {
				panic(err)
			}
			return f
		},
		false,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(doubleParameterType)
	if err != nil {
		panic(err)
	}

	byteParameterType, err := NewParameterType(
		"byte",
		FLOAT_REGEXPS,
		"int8",
		func(args ...*string) interface{} {
			f, err := transformer.Transform(*args[0], reflect.Int8)
			if err != nil {
				panic(err)
			}
			return f
		},
		false,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(byteParameterType)
	if err != nil {
		panic(err)
	}

	shortParameterType, err := NewParameterType(
		"short",
		FLOAT_REGEXPS,
		"int16",
		func(args ...*string) interface{} {
			f, err := transformer.Transform(*args[0], reflect.Int16)
			if err != nil {
				panic(err)
			}
			return f
		},
		false,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(shortParameterType)
	if err != nil {
		panic(err)
	}

	longParameterType, err := NewParameterType(
		"long",
		FLOAT_REGEXPS,
		"int32",
		func(args ...*string) interface{} {
			f, err := transformer.Transform(*args[0], reflect.Int64)
			if err != nil {
				panic(err)
			}
			return f
		},
		false,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(longParameterType)
	if err != nil {
		panic(err)
	}

	bigdecimalParameterType, err := NewParameterType(
		"bigdecimal",
		FLOAT_REGEXPS,
		"BigFloat",
		func(args ...*string) interface{} {
			f, err := transformer.Transform(*args[0], BigFloatKind)
			if err != nil {
				panic(err)
			}
			return f
		},
		false,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(bigdecimalParameterType)
	if err != nil {
		panic(err)
	}

	wordParameterType, err := NewParameterType(
		"word",
		WORD_REGEXPS,
		"string",
		func(args ...*string) interface{} {
			i, err := transformer.Transform(*args[0], reflect.String)
			if err != nil {
				panic(err)
			}
			return i
		},
		false,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(wordParameterType)
	if err != nil {
		panic(err)
	}

	stringParameterType, err := NewParameterType(
		"string",
		STRING_REGEXPS,
		"string",
		func(args ...*string) interface{} {
			matched := func(args []*string) string {
				var value string
				if args[0] == nil && args[1] != nil {
					value = *args[1]
				} else {
					value = *args[0]
				}
				return value
			}
			value := matched(args)
			value = strings.ReplaceAll(value, "\\\"", "\"")
			value = strings.ReplaceAll(value, "\\'", "'")
			i, err := transformer.Transform(value, reflect.String)
			if err != nil {
				panic(err)
			}
			return i
		},
		true,
		false,
		false,
	)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(stringParameterType)
	if err != nil {
		panic(err)
	}

	anonymousParameterType, err := createAnonymousParameterType(ANONYMOUS_REGEXPS)
	if err != nil {
		panic(err)
	}
	err = result.DefineParameterType(anonymousParameterType)
	if err != nil {
		panic(err)
	}

	return result
}

func (p *ParameterTypeRegistry) ParameterTypes() []*ParameterType {
	result := make([]*ParameterType, len(p.parameterTypeByName))
	index := 0
	for _, parameterType := range p.parameterTypeByName {
		result[index] = parameterType
		index++
	}
	return result
}

func (p *ParameterTypeRegistry) LookupByTypeName(name string) *ParameterType {
	return p.parameterTypeByName[name]
}

func (p *ParameterTypeRegistry) LookupByRegexp(parameterTypeRegexp string, expressionRegexp string, text string) (*ParameterType, error) {
	parameterTypes, ok := p.parameterTypesByRegexp[parameterTypeRegexp]
	if !ok {
		return nil, nil
	}
	if len(parameterTypes) > 1 && !parameterTypes[0].PreferForRegexpMatch() {
		generatedExpressions := NewCucumberExpressionGenerator(p).GenerateExpressions(text)
		return nil, NewAmbiguousParameterTypeError(parameterTypeRegexp, expressionRegexp, parameterTypes, generatedExpressions)
	}
	return parameterTypes[0], nil
}

func (p *ParameterTypeRegistry) DefineParameterType(parameterType *ParameterType) error {
	if _, ok := p.parameterTypeByName[parameterType.Name()]; ok {
		if len(parameterType.Name()) == 0 {
			return fmt.Errorf("The anonymous parameter type has already been defined")
		}
		return fmt.Errorf("There is already a parameter type with name %s", parameterType.Name())
	}
	p.parameterTypeByName[parameterType.Name()] = parameterType
	for _, parameterTypeRegexp := range parameterType.Regexps() {
		if _, ok := p.parameterTypesByRegexp[parameterTypeRegexp.String()]; !ok {
			p.parameterTypesByRegexp[parameterTypeRegexp.String()] = []*ParameterType{}
		}
		parameterTypes := p.parameterTypesByRegexp[parameterTypeRegexp.String()]
		if len(parameterTypes) > 0 && parameterTypes[0].PreferForRegexpMatch() && parameterType.PreferForRegexpMatch() {
			return fmt.Errorf("There can only be one preferential parameter type per regexp. The regexp /%s/ is used for two preferential parameter types, {%s} and {%s}", parameterTypeRegexp.String(), parameterTypes[0].Name(), parameterType.Name())
		}
		parameterTypes = append(parameterTypes, parameterType)
		sort.Slice(parameterTypes, func(i int, j int) bool {
			return CompareParameterTypes(parameterTypes[i], parameterTypes[j]) <= 0
		})
		p.parameterTypesByRegexp[parameterTypeRegexp.String()] = parameterTypes
	}
	return nil
}

File: python/cucumber_expressions/parameter_type_registry.py
import functools
from typing import Optional

from cucumber_expressions.parameter_type import ParameterType
from cucumber_expressions.errors import (
    CucumberExpressionError,
    AmbiguousParameterTypeError,
)

INTEGER_REGEXPS = [r"-?\d+", r"\d+"]
FLOAT_REGEXP = r"(?=.*\d.*)[-+]?\d*(?:\.(?=\d.*))?\d*(?:\d+[E][+-]?\d+)?"
WORD_REGEXP = r"[^\s]+"
STRING_REGEXPS = [r'"([^\"\\]*(\\.[^\"\\]*)*)"', r"'([^'\\]*(\\.[^'\\]*)*)'"]
ANONYMOUS_REGEXP = r".*"


class ParameterTypeRegistry:
    def __init__(self):
        self.parameter_type_by_name = {}
        self.parameter_types_by_regexp = {}
        self.define_parameter_type(
            ParameterType(
                "int", INTEGER_REGEXPS, int, lambda s=None: s and int(s), True, True
            )
        )
        self.define_parameter_type(
            ParameterType(
                "float", FLOAT_REGEXP, float, lambda s=None: s and float(s), True, False
            )
        )
        self.define_parameter_type(
            ParameterType("word", WORD_REGEXP, str, lambda s=None: s, False, False)
        )
        self.define_parameter_type(
            ParameterType(
                "string",
                STRING_REGEXPS,
                str,
                lambda s1, s2: str(s2 if s1 is None else s1)
                .replace('\\"', '"')
                .replace("\\'", "'"),
                True,
                False,
            )
        )
        self.define_parameter_type(
            ParameterType("", ANONYMOUS_REGEXP, int, lambda s=None: s, False, True)
        )

    @property
    def parameter_types(self):
        return list(self.parameter_type_by_name.values())

    def lookup_by_type_name(self, name: str) -> Optional[ParameterType]:
        return self.parameter_type_by_name.get(name)

    def lookup_by_regexp(
        self, parameter_type_regexp: str, expression_regexp, text: str
    ):
        raw_regex = rf"{parameter_type_regexp}"
        parameter_types = self.parameter_types_by_regexp.get(raw_regex)
        if not parameter_types:
            return None
        if len(parameter_types) > 1 and not parameter_types[0].prefer_for_regexp_match:
            from cucumber_expressions.cucumber_expression_generator import (
                CucumberExpressionGenerator,
            )

            generated_expressions = CucumberExpressionGenerator(
                self
            ).generate_expressions(text)
            raise AmbiguousParameterTypeError(
                parameter_type_regexp,
                expression_regexp,
                parameter_types,
                generated_expressions,
            )
        return parameter_types[0]

    def define_parameter_type(
        self, parameter_type: ParameterType
    ) -> list[ParameterType]:
        if parameter_type.name is not None:
            if parameter_type.name in self.parameter_type_by_name:
                if not len(parameter_type.name):
                    raise CucumberExpressionError(
                        "The anonymous parameter type has already been defined"
                    )
                else:
                    raise CucumberExpressionError(
                        f"There is already a parameter with name {parameter_type.name}"
                    )
            self.parameter_type_by_name[parameter_type.name] = parameter_type

        parameter_types: list[ParameterType] = []
        for parameter_type_regexp in parameter_type.regexps:
            if parameter_type_regexp not in self.parameter_types_by_regexp:
                self.parameter_types_by_regexp[parameter_type_regexp] = []
            parameter_types = self.parameter_types_by_regexp[parameter_type_regexp]
            if (
                parameter_types
                and parameter_types[0].prefer_for_regexp_match
                and parameter_type.prefer_for_regexp_match
            ):
                raise CucumberExpressionError(
                    f"""
                    There can only be one preferential parameter type per regexp.
                    The regexp '{parameter_type_regexp}' is used for two preferential parameter types,
                    {parameter_types[0].name} and {parameter_type.name}
                    """
                )
            parameter_types.append(parameter_type)
            parameter_types.sort(key=functools.cmp_to_key(ParameterType.compare))
            self.parameter_types_by_regexp[parameter_type_regexp] = parameter_types
